File: HutHelper/Controllers/NewWeb/BaseWebViewController.m
//
//  BaseWebViewController.m
//  HutHelper
//
//  Created by 张驰 on 2019/11/14.
//  Copyright © 2019 nine. All rights reserved.
//

#import "BaseWebViewController.h"
 
#import "MBProgressHUD.h"
#import "MBProgressHUD+MJ.h"
#import "HutHelper-Swift.h"
@interface BaseWebViewController ()
@property (weak, nonatomic) IBOutlet UIWebView *webView;

@end

@implementation BaseWebViewController{
     UIButton *mainAndSearchBtn;
}
- (void)viewDidLoad {
    [super viewDidLoad];
    self.view.backgroundColor = [UIColor whiteColor];
    self.navigation_item.title = _centerTitle;
    NSURL *url                = [[NSURL alloc]initWithString:_url];
    [self.navigationController.navigationBar setTitleTextAttributes:@{NSForegroundColorAttributeName:[UIColor blackColor]}];
    [self customBackButton];
    _webView.delegate=self;
    [_webView loadRequest:[NSURLRequest requestWithURL:url]];
}

// 自定义返回按钮
- (void)customBackButton{
    /**按钮*/
    UIView *rightButtonView = [[UIView alloc] initWithFrame:CGRectMake(-20,0, 40, 40)];
    
    mainAndSearchBtn = [[UIButton alloc] initWithFrame:CGRectMake(-20,0, 40, 40)];
    [rightButtonView addSubview:mainAndSearchBtn];
    [mainAndSearchBtn setImage:[UIImage imageNamed:@"ico_menu_back"] forState:UIControlStateNormal];
    [mainAndSearchBtn addTarget:self action:@selector(backBtnClicked:) forControlEvents:UIControlEventTouchUpInside];
    UIBarButtonItem *rightCunstomButtonView = [[UIBarButtonItem alloc] initWithCustomView:rightButtonView];
    self.navigation_item.leftBarButtonItem = rightCunstomButtonView;
}
// 返回按钮按下
- (void)backBtnClicked:(UIButton *)sender{
    if(self.webView.canGoBack){
        [self.webView goBack];
    }else{
        [self.navigationController popViewControllerAnimated:YES];
    }
}
- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    self.navigationItem.backBarButtonItem = mainAndSearchBtn;
    self.navigationController.interactivePopGestureRecognizer.delegate = self;
}
- (void)viewWillDisappear:(BOOL)animated
{
    
    [super viewWillDisappear:animated];
}
/** webView的代理方法*/

- (void)webViewDidStartLoad:(UIWebView *)webView
{
    //提示用户正在加载
    [MBProgressHUD showMessage:@"加载中" toView:self.view];
}

- (void)webViewDidFinishLoad:(UIWebView *)webView
{
    
    HideAllHUD
    
}

- (void)webView:(UIWebView *)webView didFailLoadWithError:(NSError *)error
{
    //隐藏显示
    HideAllHUD
    [MBProgressHUD showError:@"网络错误" toView:self.view];
}

@end

File: HutHelper/Controllers/NewWeb/BaseWebViewController.swift
//
//  BaseWebViewController.swift
//  HutHelperSwift
//
//  Created by 张驰 on 2020/1/11.
//  Copyright © 2020 张驰. All rights reserved.
//

import UIKit
import WebKit
import RxCocoa
import RxSwift

class BaseWebController: UIViewController{
    var webURL = ""
    var navTitle = ""
    
    // 左边返回
    private lazy var leftBarButton:UIButton = {
        let btn = UIButton.init(type: .custom)
            btn.frame = CGRect(x:-5, y:0, width:20, height: 30)
            btn.setImage(UIImage(named: "ico_menu_back"), for: .normal)
            btn.rx.tap.subscribe(onNext:{[weak self] in
            self?.navigationController?.popViewController(animated: true)
        })
        return btn
    }()
    
    convenience init(webURL:String,navTitle:String){
        self.init()
        self.webURL = webURL
        self.navTitle = navTitle
    }
    //web页面
    lazy private var webview: WKWebView = {
        self.webview = WKWebView.init(frame: self.view.bounds)
        self.webview.uiDelegate = self
        self.webview.navigationDelegate = self
        return self.webview
    }()
    // 进度条
    lazy private var progressView: UIProgressView = {
        self.progressView = UIProgressView.init(frame: CGRect(x: CGFloat(0), y: CGFloat(65), width: UIScreen.main.bounds.width, height: 10))
        self.progressView.tintColor = UIColor.green      // 进度条颜色
        self.progressView.trackTintColor = UIColor.white // 进度条背景色
        return self.progressView
    }()
    
    override func viewDidLoad() {
        super.viewDidLoad()

        configUI()
        configWeb()
    }
    func configWeb(){
        webview.addObserver(self, forKeyPath: "estimatedProgress", options: .new, context: nil)
        webview.load(URLRequest.init(url: URL.init(string: webURL)!))
    }
    func configUI(){
        view.addSubview(webview)
        self.navigation.item.title = navTitle
        self.navigation.item.leftBarButtonItem = UIBarButtonItem.init(customView: leftBarButton)
        self.navigation.bar.isShadowHidden = true
    }

    override func observeValue(forKeyPath keyPath: String?, of object: Any?, change: [NSKeyValueChangeKey : Any]?, context: UnsafeMutableRawPointer?) {
            
            if keyPath == "estimatedProgress"{
                progressView.alpha = 1.0
                
                DispatchQueue.main.async {
                    print(self.webview.estimatedProgress)
                    self.progressView.setProgress(Float(self.webview.estimatedProgress), animated: true)
                }

                print(webview.estimatedProgress)
                if webview.estimatedProgress >= 1.0 {
                    UIView.animate(withDuration: 0.3, delay: 0.1, options: .curveEaseOut, animations: {
                        self.progressView.alpha = 0
                    }, completion: { (finish) in
                        self.progressView.setProgress(0.0, animated: false)
                    })
                }
            }
    }
    deinit {
            webview.removeObserver(self, forKeyPath: "estimatedProgress")
            webview.uiDelegate = nil
            webview.navigationDelegate = nil
    }

}

extension BaseWebController:WKUIDelegate,WKNavigationDelegate  {
    
    func webView(_ webView: WKWebView, didStartProvisionalNavigation navigation: WKNavigation!) {
            let hud = MBProgressHUD.showAdded(to: self.view, animated: true)
            hud.label.text = "加载中"
    }
        
    func webView(_ webView: WKWebView, didCommit navigation: WKNavigation!) {
            print("开始获取网页内容")
    }
        
    func webView(_ webView: WKWebView, didFinish navigation: WKNavigation!) {
            MBProgressHUD.hideAllHUDs(for: self.view, animated: true)
    }
        
    func webView(_ webView: WKWebView, didFail navigation: WKNavigation!, withError error: Error) {
            print("加载失败")
    }
        
    func webView(_ webView: WKWebView, decidePolicyFor navigationAction: WKNavigationAction, decisionHandler: @escaping (WKNavigationActionPolicy) -> Void) {
            decisionHandler(.allow);
    }
}
